### PR TITLE
Add Anomali Threatstream Analyzer

### PR DESCRIPTION
**Pull request by BenjaminH-Sogeti**
# Description
Added 3 Analyzer for Anomali Threastream tools. 

- Intelligence API. Threat Intelligence base API
- Confidence API. Information of "maliciously confidence observable"
- PassiveDSN API. PassiveDNS from anomali data 

## Related issues
#966 

## Type of change
- [x] New feature (non-breaking change which adds functionality).

# Checklist

- [x] I have read and understood the rules about [how to Contribute](https://intelowl.readthedocs.io/en/latest/Contribute.html) to this project
- [x] The pull request is for the branch `develop`
- [x] A new analyzer or connector was added, in which case:
    - [x] [Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Usage.md) file was updated.
    - [ ] [Advanced-Usage](./Advanced-Usage.md) was updated (in case the analyzer/connector provides additional optional configuration).
    - [x] Secrets were added in [env_file_app_template](https://github.com/intelowlproject/IntelOwl/blob/master/docker/env_file_app_template), [env_file_app_ci](https://github.com/certego/IntelOwl/blob/master/docker/env_file_app_ci) and in the [Installation](./Installation.md) docs, if necessary.
    - [ ] If the analyzer/connector requires mocked testing, `_monkeypatch()` was used in it's class to apply the necessary decorators.
    - [ ] If a File analyzer was added, it's name was explicitly defined in [test_file_scripts.py](https://github.com/intelowlproject/IntelOwl/blob/master/tests/analyzers_manager/test_file_scripts.py) (not required for Observable Analyzers).
- [ ] If external libraries/packages with restrictive licenses were used, they were added in the [Legal Notice](https://github.com/certego/IntelOwl/blob/master/.github/legal_notice.md) section.
- [ ] The tests gave 0 errors.
- [x] Linters (`Black`, `Flake`, `Isort`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [x] The commits were squashed into a single one (optional, they will be squashed anyway by the maintainer)
  
### Important Rules
- If your changes decrease the overall tests coverage (you will know after the Codecov CI job is done), you should add the required tests to fix the problem
- Every time you make changes to the PR and you think the work is done, you should explicitly ask for a review

# Real World Example
### Without filter 
image
6 entries returned
### With active filter 
image
Only 3 active entries 
### With limit filter
image
Return 4 object out of the 6 existing
### With date filter
image
Return no object since the date is in the future
## Domain analysis
image
### Confidence 
{
  "max_confidence": 76,
  "min_confidence": 76,
  "average_confidence": 76
}

### PassiveDNS
{
  "cached": true,
  "results": [
    {
      "ip": "155.94.169.48",
      "domain": "e00v3v3m.cn",
      "rrtype": "A",
      "source": "VirusTotal",
      "last_seen": "2022-04-12 15:06:12",
      "first_seen": null
    }
  ],
  "success": true
}

### Intelligence
{
  "meta": {
    "next": null,
    "took": 182,
    "limit": 100,
    "offset": 0,
    "total_count": 3
  },
  "objects": [
    {
      "id": 58666078226,
      "ip": null,
      "asn": "",
      "org": "",
      "tlp": null,
      "meta": {
        "detail2": "imported by user 27067",
        "severity": "high",
        "registrant_name": "练巧芳",
        "registrant_email": "ibvoxpxdiec@hotmail.com",
        "registration_created": "2021-06-13T13:40:51"
      },
      "rdns": null,
      "sort": [
        1649778921986,
        "58666078226"
      ],
      "tags": [
        {
          "id": "rtt",
          "name": "#jaccs"
        },
        {
          "id": "er1",
          "name": "#jaccscard"
        },
        {
          "id": "l0c",
          "name": "#phishing"
        },
        {
          "id": "k1s",
          "name": "https://twitter.com/KesaGataMe0/status/1513906024021106693"
        },
        {
          "id": "qh4",
          "name": "KesaGataMe0"
        }
      ],
      "type": "url",
      "uuid": "adccd58e-0392-4053-955b-7c0869969ba4",
      "itype": "phish_url",
      "value": "https://www.jacss.co.jp-service-tranid-00001-0001n.e00v3v3m.cn/",
      "source": "FirstEnergy",
      "status": "active",
      "country": null,
      "feed_id": 0,
      "subtype": null,
      "latitude": null,
      "is_public": false,
      "longitude": null,
      "update_id": 14184323061,
      "confidence": 85,
      "created_ts": "2022-04-12T15:53:34.321Z",
      "workgroups": [],
      "description": null,
      "is_editable": false,
      "modified_ts": "2022-04-12T15:55:21.986Z",
      "threat_type": "phish",
      "threatscore": 68,
      "is_anonymous": false,
      "resource_uri": "/api/v2/intelligence/58666078226/",
      "expiration_ts": "2022-07-11T15:53:32.260Z",
      "source_created": null,
      "source_modified": null,
      "import_session_id": 5125573,
      "retina_confidence": 85,
      "trusted_circle_ids": [
        10978
      ],
      "can_add_public_tags": true,
      "owner_organization_id": 2316,
      "source_reported_confidence": 100
    },
    {
      "id": 58666078206,
      "ip": null,
      "asn": "",
      "org": "",
      "tlp": null,
      "meta": {
        "detail2": "imported by user 27067",
        "severity": "high",
        "registrant_name": "练巧芳",
        "registrant_email": "ibvoxpxdiec@hotmail.com",
        "registration_created": "2021-06-13T13:40:51"
      },
      "rdns": null,
      "sort": [
        1649778921952,
        "58666078206"
      ],
      "tags": [
        {
          "id": "v6j",
          "name": "#jaccs"
        },
        {
          "id": "koz",
          "name": "#jaccscard"
        },
        {
          "id": "93o",
          "name": "#phishing"
        },
        {
          "id": "7e6",
          "name": "https://twitter.com/KesaGataMe0/status/1513906024021106693"
        },
        {
          "id": "2v0",
          "name": "KesaGataMe0"
        }
      ],
      "type": "domain",
      "uuid": "45cb9e76-4e09-4445-971d-6ff83c570bfc",
      "itype": "phish_domain",
      "value": "www.jacss.co.jp-service-tranid-00001-0001n.e00v3v3m.cn",
      "source": "FirstEnergy",
      "status": "active",
      "country": null,
      "feed_id": 0,
      "subtype": null,
      "latitude": null,
      "is_public": false,
      "longitude": null,
      "update_id": 14184323058,
      "confidence": 85,
      "created_ts": "2022-04-12T15:53:34.078Z",
      "workgroups": [],
      "description": null,
      "is_editable": false,
      "modified_ts": "2022-04-12T15:55:21.952Z",
      "threat_type": "phish",
      "threatscore": 68,
      "is_anonymous": false,
      "resource_uri": "/api/v2/intelligence/58666078206/",
      "expiration_ts": "2022-07-11T15:53:32.260Z",
      "source_created": null,
      "source_modified": null,
      "import_session_id": 5125573,
      "retina_confidence": 85,
      "trusted_circle_ids": [
        10978
      ],
      "can_add_public_tags": true,
      "owner_organization_id": 2316,
      "source_reported_confidence": 100
    },
    {
      "id": 58666078221,
      "ip": "76.164.203.88",
      "asn": "36114",
      "org": "Versaweb, LLC",
      "tlp": null,
      "meta": {
        "detail2": "imported by user 27067",
        "severity": "high",
        "registrant_name": "练巧芳",
        "registrant_email": "ibvoxpxdiec@hotmail.com",
        "registration_created": "2021-06-13T13:40:51"
      },
      "rdns": null,
      "sort": [
        1649778843281,
        "58666078221"
      ],
      "tags": [
        {
          "id": "bcr",
          "name": "#jaccs"
        },
        {
          "id": "9ne",
          "name": "#jaccscard"
        },
        {
          "id": "wh6",
          "name": "#phishing"
        },
        {
          "id": "6zl",
          "name": "https://twitter.com/KesaGataMe0/status/1513906024021106693"
        },
        {
          "id": "t2b",
          "name": "KesaGataMe0"
        }
      ],
      "type": "domain",
      "uuid": "c8263b88-ee36-41f8-84d5-48ecf0e100cb",
      "itype": "phish_domain",
      "value": "e00v3v3m.cn",
      "source": "FirstEnergy",
      "status": "active",
      "country": "US",
      "feed_id": 0,
      "subtype": null,
      "latitude": 36.122,
      "is_public": false,
      "longitude": -115.086,
      "update_id": 14184321382,
      "confidence": 76,
      "created_ts": "2022-04-12T15:53:34.261Z",
      "workgroups": [],
      "description": null,
      "is_editable": false,
      "modified_ts": "2022-04-12T15:54:03.281Z",
      "threat_type": "phish",
      "threatscore": 61,
      "is_anonymous": false,
      "resource_uri": "/api/v2/intelligence/58666078221/",
      "expiration_ts": "2022-07-11T15:53:32.260Z",
      "source_created": null,
      "source_modified": null,
      "import_session_id": 5125573,
      "retina_confidence": 76,
      "trusted_circle_ids": [
        10978
      ],
      "can_add_public_tags": true,
      "owner_organization_id": 2316,
      "source_reported_confidence": 100
    }
  ]
}

### CODE REVIEW

**Comment by BenjaminH-Sogeti** [comment]
Tests are giving 3 errors, but exactly the same as these on already raised on develop branch. (or I don't run them correctly)

**Comment by mlodic** [comment]
Consider that was your first PR I needed to approve the tests. They are running now

File: configuration/analyzer_config.json [hunk 1]
```diff
@@ -42,6 +42,134 @@
...
+      },
+      "api_key_name": {
+        "env_var_key": "ANOMALI_THREATSTREAM_API_KEY",
+        "type": "string",
+        "required": true,
+        "default": null,
+        "description": "API Key for Anomali Threatstream"
+      }
+    },
+	"params": {
```
Comment (mlodic) [inline configuration/analyzer_config.json:74]: this is not correctly aligned, same in the other 2 definitions

File: api_app/analyzers_manager/observable_analyzers/threatstream.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+                "modified_ts__gte": self.modified_after,
+            }  # If value = None don't enter in filter
+            uri = "v2/intelligence/"
+        elif self.analysis_type == "confidence":
+            params = {"type": "confidence", "value": self.observable_name}
+            uri = "v1/inteldetails/confidence_trend/"
+        elif self.analysis_type == "passive_dns":
+            if self.observable_classification == self.ObservableTypes.IP:
+                uri = f"v1/pdns/ip/{self.observable_name}"
+            if self.observable_classification == self.ObservableTypes.DOMAIN:
```
Comment (mlodic) [inline api_app/analyzers_manager/observable_analyzers/threatstream.py:41]: elif + we should add an else with AnalyzerConfigurationException cause observable type not supported

**Review by mlodic — CHANGES_REQUESTED** [review]
thanks, good first contributions. Just some adjusts and we are done

File: docs/source/Usage.md [hunk 3]
```diff
@@ -276,6 +275,10 @@ The following is the list of the available analyzers you can run out-of-the-box.
 * `Spyse`: Scan domains, IPs, emails and CVEs using Spyse's API. Register [here](https://spyse.com/user/registration).
 * `FileScan_Search`: Finds reports and uploaded files by various tokens, like hash, filename, verdict, IOCs etc via [FileScan.io  API](https://www.filescan.io/api/docs).
 * `Virushee_CheckHash`: Search for a previous analysis of a file by its hash (SHA256/SHA1/MD5) on [Virushee API](https://api.virushee.com/).
+* `Anomali_Threatstream_Confidence`: Give max, average and min confidence for an observable. On [Anomali Threatstream](https://www.anomali.com/products/threatstream) Confidence API.
```
[the remark this answers is not in the document]
Comment (BenjaminH-Sogeti) [reply docs/source/Usage.md]: Done, only kept PassiveDNS in it

File: api_app/analyzers_manager/observable_analyzers/threatstream.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            if self.must_active:
+                self.active = "active"
+            params = {
+                "value__contains": self.observable_name,
+                "limit": self.limit,
+                "status": self.active,
+                "confidence__gt": self.minimal_confidence,
+                "modified_ts__gte": self.modified_after,
+            }  # If value = None don't enter in filter
+            uri = "v2/intelligence/"
```
[the remark this answers is not in the document]
Comment (BenjaminH-Sogeti) [reply api_app/analyzers_manager/observable_analyzers/threatstream.py:35]: see slack PM

**Pull request merged**